psi4/src/psi4/libfock/DFJLinK.cc

```diff
@@ -0,0 +1,932 @@
+
+            if (do_J_) {
+                prev_J_ao_.push_back(std::make_shared<Matrix>("J Prev", J_ao_[N]->nrow(), J_ao_[N]->ncol()));
+                delta_J_ao_.push_back(std::make_shared<Matrix>("Delta J", J_ao_[N]->nrow(), J_ao_[N]->ncol()));
+            }
+        
+            if (do_K_) {
+                prev_K_ao_.push_back(std::make_shared<Matrix>("K Prev", K_ao_[N]->nrow(), K_ao_[N]->ncol()));
+                delta_K_ao_.push_back(std::make_shared<Matrix>("Delta K", K_ao_[N]->nrow(), K_ao_[N]->ncol()));
+            }
+        }
+    } else {
+        for (size_t N = 0; N < D_ao_.size(); N++) {
+            delta_D_ao_[N]->copy(D_ao_[N]);
+            delta_D_ao_[N]->subtract(prev_D_ao_[N]);
+        }
+    }
+}
+void DFJLinK::incfock_postiter() {
+    if (do_incfock_iter_) {
+        for (size_t N = 0; N < D_ao_.size(); N++) {
+
+            if (do_J_) {
+                prev_J_ao_[N]->add(delta_J_ao_[N]);
+                J_ao_[N]->copy(prev_J_ao_[N]);
+            }
+
+            if (do_K_) {
+                prev_K_ao_[N]->add(delta_K_ao_[N]);
+                K_ao_[N]->copy(prev_K_ao_[N]);
+            }
+
+            prev_D_ao_[N]->copy(D_ao_[N]);
+        }
+    } else {
+        for (size_t N = 0; N < D_ao_.size(); N++) {
+            if (do_J_) prev_J_ao_[N]->copy(J_ao_[N]);
+            if (do_K_) prev_K_ao_[N]->copy(K_ao_[N]);
+            prev_D_ao_[N]->copy(D_ao_[N]);
+        }
+    }
+}
+
+void DFJLinK::compute_JK() {
+  
+    // wK not supported in DFJLinK yet
+    if (do_wK_) throw PSIEXCEPTION("LINK does not support wK integrals yet!");
+   
+    // => Set up Incremental Fock and Density Screening if required <= //
+    
+    // explicit setup of Incfock for this SCF iteration
+    if (incfock_) {
+        timer_on("DFJLinK: INCFOCK Preprocessing");
+        
+	incfock_setup();
+
+        int reset = options_.get_int("INCFOCK_FULL_FOCK_EVERY");
+        double incfock_conv = options_.get_double("INCFOCK_CONVERGENCE");
+        double Dnorm = Process::environment.globals["SCF D NORM"];
+        // Do IFB on this iteration?
+        do_incfock_iter_ = (Dnorm >= incfock_conv) && !initial_iteration_ && (incfock_count_ % reset != reset - 1);
+        
+        if (!initial_iteration_ && (Dnorm >= incfock_conv)) incfock_count_ += 1;
+        
+	timer_off("DFJLinK: INCFOCK Preprocessing");
+    }
+    
+    // define matrices to be used for J/K construction
+    std::vector<SharedMatrix>& D_ref = (do_incfock_iter_ ? delta_D_ao_ : D_ao_);
+    std::vector<SharedMatrix>& J_ref = (do_incfock_iter_ ? delta_J_ao_ : J_ao_);
+    std::vector<SharedMatrix>& K_ref = (do_incfock_iter_ ? delta_K_ao_ : K_ao_);
+
+    // update ERI engine density matrices for density screening
+    if (density_screening_) {
+        for (auto eri_computer : eri_computers_["4-Center"]) {
+            eri_computer->update_density(D_ref);
+	}
+    }
+
+    // => Perform matrix calculations <= //
+    
+    // Direct DF-J
+    if (do_J_) {
```

Comment:
if you haven't already, it might be a useful sanity check to `set_do_J(false)` and same for K. Just a run -- no need for a test, imo.

Reply:
Good point! I will do this.

Reply:
Interesting... adding set_do_J(false) to the constructor causes a segfault, but adding set_do_K(false) to the constructor does not. I shall explore this further.

Reply:
Agreed on defer -- sounds like a universal weakness and probably not one likely to be hit. Thanks for investigating.